### PR TITLE
[file_selector] add getDirectoryPaths implementation on Linux

### PR DESCRIPTION
**Pull request by adpinola**
This PR adds the Linux implementation for retrieving multiple directories paths from a select folder dialog.

Issue: [Support for selection of multiple directories, through desktop's native open panel, in 'file_selector' package #74323](https://github.com/flutter/flutter/issues/74323)

VirtualBoxVM_rwSR7C6lcO
_New option on example application_

image
_GetDirectoriesPaths_

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/file_selector/file_selector_linux/pubspec.yaml [hunk 1]
```diff
@@ -2,7 +2,11 @@ name: file_selector_linux
 description: Liunx implementation of the file_selector plugin.
 repository: https://github.com/flutter/plugins/tree/main/packages/file_selector/file_selector_linux
 issue_tracker: https://github.com/flutter/flutter/issues?q=is%3Aissue+is%3Aopen+label%3A%22p%3A+file_selector%22
-version: 0.9.0+1
+version: 0.9.0+2
+
+# TODO(adpinola): remove this once file_selector_platform_interface version is updated to 2.3.0
+# with getDirectoriesPaths method.
+publish_to: none
```
Comment (stuartmorgan) [inline packages/file_selector/file_selector_linux/pubspec.yaml:9]: Please don't disable checks by marking things as non-published; the failures from the temporary use of paths are good, and should not be bypassed.

File: packages/file_selector/file_selector_linux/lib/file_selector_linux.dart [hunk 2]
```diff
@@ -16,6 +16,7 @@ const String _typeGroupMimeTypesKey = 'mimeTypes';
 const String _openFileMethod = 'openFile';
 const String _getSavePathMethod = 'getSavePath';
 const String _getDirectoryPathMethod = 'getDirectoryPath';
+const String _getDirectoriesPathsMethod = 'getDirectoriesPaths';
```
Comment (stuartmorgan) [inline packages/file_selector/file_selector_linux/lib/file_selector_linux.dart:19]: This should not use a new method. Please follow the same structure as `openFile`.

File: packages/file_selector/file_selector_linux/linux/file_selector_plugin.cc [hunk 3]
```diff
@@ -192,7 +196,8 @@ static void method_call_cb(FlMethodChannel* channel, FlMethodCall* method_call,
   FlValue* args = fl_method_call_get_args(method_call);
 
   g_autoptr(FlMethodResponse) response = nullptr;
-  if (strcmp(method, kOpenFileMethod) == 0) {
+  if (strcmp(method, kOpenFileMethod) == 0 ||
+      strcmp(method, kGetDirectoriesPathsMethod) == 0) {
```
Comment (stuartmorgan) [inline packages/file_selector/file_selector_linux/linux/file_selector_plugin.cc:200]: The logic to extract one value from a list of returns should be on the Dart side, as it is for files.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Comment by adpinola** [comment]
Changes applied, thanks

File: packages/file_selector/file_selector_linux/.gitignore [hunk 4]
```diff
@@ -3,3 +3,7 @@
 .flutter-plugins
 .flutter-plugins-dependencies
 pubspec.lock
+
+# Linux specific
```
Comment (stuartmorgan) [inline packages/file_selector/file_selector_linux/.gitignore:7]: Linux-specific ignores belong in `linux/.gitignore`, not the top-level `.gitignore`.

Why are these needed though? What is creating them? I've never seen these locally.

File: packages/file_selector/file_selector_linux/example/lib/get_multiple_directories_page.dart [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      confirmButtonText: confirmButtonText,
+    );
+    if (directoryPaths.isEmpty) {
+      // Operation was canceled by the user.
+      return;
+    }
+    String paths = '';
+    for (final String? path in directoryPaths) {
+      paths += '${path!} \n';
+    }
```
Comment (stuartmorgan) [inline packages/file_selector/file_selector_linux/example/lib/get_multiple_directories_page.dart:27]: Why not just `directoryPaths.join('\n')`?

Comment (adpinola) [reply packages/file_selector/file_selector_linux/example/lib/get_multiple_directories_page.dart]: much better, for sure

File: packages/file_selector/file_selector_linux/.gitignore (hunk 4 above)
Comment (adpinola) [reply packages/file_selector/file_selector_linux/.gitignore]: Those are being created when running the example app in linux

**Comment by stuartmorgan** [comment]
What's the state of this PR? Is it ready for re-review?

**Comment by adpinola** [comment]
> What's the state of this PR? Is it ready for re-review?
Yes, it is.

**Review by stuartmorgan — APPROVED** [review]
LGTM

@cbracken for second review.

File: packages/file_selector/file_selector_linux/linux/.gitignore [hunk 6]
```diff
@@ -0,0 +1,2 @@
+CMakeCache.txt
+CMakeFiles/
```
Comment (cbracken) [inline packages/file_selector/file_selector_linux/linux/.gitignore:2]: Extremely pedantic nitpick: add a trailing newline. :)

**Review by cbracken — APPROVED** [review]
LGTM stamp from a Japanese personal seal

**Comment by stuartmorgan** [comment]
I'm going to land this as-is so we're not doing a new test run just for the newline; if we ever need to touch that file again we can always add the missing newline then.

**Comment by stuartmorgan** [comment]
`submit-queue` status is stale, so landing manually.
